src/plugins/intel_cpu/src/nodes/scaled_attn.cpp

```diff
@@ -806,6 +824,288 @@ bool ScaledDotProductAttention::isSupportedOperation(const std::shared_ptr<const
+    auto mem_desc = std::make_shared<CpuBlockedMemoryDesc>(ov::element::i32,
+        Shape(new_shape),
+        new_shape,
+        VectorDims{0, 1},
+        0,
+        VectorDims{},
+        hidden_state_k->getDescWithType<BlockedMemoryDesc>()->getStrides());
+    hidden_state_k->redefineDesc(mem_desc);
+    hidden_state_v->redefineDesc(mem_desc);
+
+    if (!beam_table_k) {
+        beam_table_k.reset(hidden_state_k);
+        beam_table_v.reset(hidden_state_v);
+    }
+
+    // first token
+    if (L0 == 0 || is_reset) {
+        for (size_t b = 0; b < B; b++) {
+            for (size_t l = 0; l < L0 + L1; l++) {
+                beam_table_k.at<int32_t>({b, l}) = b;
+                beam_table_v.at<int32_t>({b, l}) = b;
+            }
+        }
+        return;
+    }
+
+    // beam order is like [0, 1, 2,...]
+    bool no_reorder = true;
+    for (size_t i = 0; i < B; i++) {
+        if (beam_idx.data<int32_t>()[i] != static_cast<int32_t>(i)) {
+            no_reorder = false;
+            break;
+        }
+    }
+
+    // reorder
+    if (!no_reorder) {
+        m_tmp_reorder.resize<int32_t>({B, L0});
+        for (size_t i = 0; i < B; i++) {
+            std::memcpy(&m_tmp_reorder.at<int32_t>({i}),
+                        &beam_table_k.at<int32_t>({i}),
+                        sizeof(int32_t) * L0);
+        }
+        auto* table = beam_idx.data<int32_t>();
+        // beam table is same for both k,v state
+        for (size_t i = 0; i < B; i++) {
+            std::memcpy(&beam_table_k.at<int32_t>({i}),
+                        &m_tmp_reorder.at<int32_t>({static_cast<size_t>(table[i])}),
+                        sizeof(int32_t) * L0);
+            std::memcpy(&beam_table_v.at<int32_t>({i}),
+                        &m_tmp_reorder.at<int32_t>({static_cast<size_t>(table[i])}),
+                        sizeof(int32_t) * L0);
+        }
+    }
+    // second token itself
+    for (size_t i = 0; i < B; i++) {
+        for (size_t j = 0; j < L1; j++) {
+            beam_table_k.at<int32_t>({i, L0 + j}) = i;
+            beam_table_v.at<int32_t>({i, L0 + j}) = i;
+        }
+    }
+}
+
+void ScaledDotProductAttention::updatePastkv(const MemoryPtr& mem_cur_k, const MemoryPtr& mem_cur_v) {
+    std::vector<size_t> order = {0, 1, 2, 3};
+    if (!m_config.config.permute_axes.empty()) {
+        order = m_config.config.permute_axes;
+    }
+    PlainTensor cur_k, past_k;
+    PlainTensor cur_v, past_v;
+    cur_k.reset(mem_cur_k);
+    cur_v.reset(mem_cur_v);
+    cur_k = cur_k.permute(order);
+    cur_v = cur_v.permute(order);
+    auto B = cur_k.size(0);
+    auto H = cur_k.size(1);
+    auto L1 = cur_k.size(2);
+    auto S = cur_k.size(3);
+    size_t L0 = 0;
+    auto& reverse_order = m_config.reverse_order;
+    auto reverse = [&reverse_order] (const std::vector<size_t>& cur) {
+        std::vector<size_t> result(cur.size());
+        for (size_t i = 0; i < cur.size(); i++) {
+            result[i] = cur[reverse_order[i]];
```

Comment:
can be simplified as `result[order[i]] = cur[i]` w/o the need for `reverse_order` and corresponding logics.

because for oshape permuted from ishape using order, we have following equation:
 `oshape[i] == ishape[order[i]]`
 
so to recover ishape from oshape, we only need to `recovered_ishape[order[i]] = oshape[i];`

Reply:
Removed.